File: archaic/src/main/java/resonantinduction/archaic/fluid/grate/TileGrate.java
package resonantinduction.archaic.fluid.grate;

import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.PriorityQueue;

import net.minecraft.block.material.Material;
import net.minecraft.client.renderer.texture.IconRegister;
import net.minecraft.util.Icon;
import net.minecraft.world.IBlockAccess;
import net.minecraftforge.common.ForgeDirection;
import net.minecraftforge.fluids.Fluid;
import net.minecraftforge.fluids.FluidContainerRegistry;
import net.minecraftforge.fluids.FluidRegistry;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.fluids.FluidTank;
import net.minecraftforge.fluids.FluidTankInfo;
import resonantinduction.core.Reference;
import resonantinduction.core.fluid.TilePressureNode;
import resonantinduction.core.grid.fluid.FluidPressureNode;
import universalelectricity.api.vector.Vector3;
import calclavia.lib.config.Config;
import calclavia.lib.prefab.tile.IRotatable;
import calclavia.lib.utility.FluidUtility;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

public class TileGrate extends TilePressureNode implements IRotatable
{
	@Config(comment = "The multiplier for the influence of the grate. This is multiplied by the pressure applied.")
	private static double grateEffectMultiplier = 10;

	@SideOnly(Side.CLIENT)
	private static Icon iconFront, iconSide;

	private GratePathfinder gratePath;

	public TileGrate()
	{
		super(Material.rock);
		isOpaqueCube = false;
		normalRender = true;
		rotationMask = Byte.parseByte("111111", 2);
		node = new FluidPressureNode(this);
		node.maxFlowRate = getPressureTank().getCapacity();
	}

	@Override
	public FluidTank getInternalTank()
	{
		if (this.tank == null)
		{
			this.tank = new FluidTank(2 * FluidContainerRegistry.BUCKET_VOLUME);
		}

		return this.tank;
	}

	@Override
	public Icon getIcon(IBlockAccess world, int side)
	{
		return side == getDirection().ordinal() ? iconFront : iconSide;
	}

	@Override
	public Icon getIcon(int side, int metadata)
	{
		return side == 1 ? iconFront : iconSide;
	}

	@Override
	@SideOnly(Side.CLIENT)
	public void registerIcons(IconRegister iconRegister)
	{
		iconFront = iconRegister.registerIcon(Reference.PREFIX + "grate_front");
		iconSide = iconRegister.registerIcon(Reference.PREFIX + "grate");
	}

	@Override
	public FluidTankInfo[] getTankInfo(ForgeDirection from)
	{
		return new FluidTankInfo[] { getPressureTank().getInfo() };
	}

	@Override
	public boolean canFill(ForgeDirection from, Fluid fluid)
	{
		return getDirection() != from;
	}

	@Override
	public boolean canDrain(ForgeDirection from, Fluid fluid)
	{
		return getDirection() != from;
	}

	@Override
	public void updateEntity()
	{
		super.updateEntity();

		if (!world().isRemote)
		{
			if (ticks % 10 == 0)
			{
				int pressure = node.getPressure(getDirection());
				int blockEffect = (int) Math.abs(pressure * grateEffectMultiplier);

				if (pressure > 0)
				{
					// Fill
					if (getPressureTank().getFluidAmount() >= FluidContainerRegistry.BUCKET_VOLUME)
					{
						if (gratePath == null)
						{
							gratePath = new GratePathfinder(true);
							gratePath.startFill(new Vector3(this), getPressureTank().getFluid().getFluid().getID());
						}

						int filledInWorld = gratePath.tryFill(getPressureTank().getFluidAmount(), blockEffect);
						getPressureTank().drain(filledInWorld, true);
					}
				}
				else if (pressure < 0)
				{
					// Drain
					int maxDrain = getPressureTank().getCapacity() - getPressureTank().getFluidAmount();
					// System.out.println(maxDrain);
					if (maxDrain > 0)
					{
						if (gratePath == null)
						{
							gratePath = new GratePathfinder(false);

							if (!gratePath.startDrain(new Vector3(this)))
							{
								resetPath();
							}
						}

						if (gratePath != null && gratePath.tryPopulateDrainMap(blockEffect))
						{
							getPressureTank().fill(gratePath.tryDrain(maxDrain, true), true);
						}
					}
				}
			}
		}
	}

	@Override
	public int fill(ForgeDirection from, FluidStack resource, boolean doFill)
	{
		return getPressureTank().fill(resource, doFill);
	}

	@Override
	public FluidStack drain(ForgeDirection from, FluidStack resource, boolean doDrain)
	{
		if (resource != null)
			return drain(from, resource.amount, doDrain);
		return null;
	}

	@Override
	public FluidStack drain(ForgeDirection from, int maxDrain, boolean doDrain)
	{
		return getPressureTank().drain(maxDrain, doDrain);
	}

	/**
	 * Pathfinding operations
	 * 
	 * @author Calclavia
	 */
	public void resetPath()
	{
		this.gratePath = null;
	}

	public class GratePathfinder
	{
		/**
		 * The starting vector for our grate.
		 */
		Vector3 start;

		/**
		 * All the back trace blocks tracing back to the original.
		 */
		HashMap<Vector3, Vector3> navigationMap = new HashMap<Vector3, Vector3>();

		/**
		 * The nodes we're currently working on.
		 */
		PriorityQueue<ComparableVector> workingNodes;

		/**
		 * The list of blocks to drain.
		 */
		PriorityQueue<ComparableVector> drainNodes = new PriorityQueue<ComparableVector>(1024, Collections.reverseOrder());

		/**
		 * The type of fluid we're dealing with. When draining, this will be the type of fluid being
		 * drained.
		 */
		public Fluid fluidType;

		public GratePathfinder(boolean checkVertical)
		{
			if (checkVertical)
			{
				this.workingNodes = new PriorityQueue<ComparableVector>();
			}
			else
			{
				this.workingNodes = new PriorityQueue<ComparableVector>(1024, new Comparator()
				{

					@Override
					public int compare(Object a, Object b)
					{
						TileGrate.ComparableVector wa = (TileGrate.ComparableVector) a;
						TileGrate.ComparableVector wb = (TileGrate.ComparableVector) b;

						return wa.iterations - wb.iterations;
					}
				});
			}
		}

		/**
		 * Traces back to the start position to see if this fluid position is connected with the
		 * starting position through fluid mediums.
		 */
		public boolean isConnected(Vector3 check)
		{
			if (check.equals(this.start))
				return true;
			do
			{
				check = this.navigationMap.get(check);

				if (check == null)
					return false;

				if (check.equals(this.start))
					return true;
			}
			while (FluidUtility.getFluidFromBlock(TileGrate.this.worldObj, check) != null && fluidType.getID() == FluidUtility.getFluidFromBlock(TileGrate.this.worldObj, check).getID());

			return false;
		}

		public void startFill(Vector3 start, int fluidID)
		{
			this.fluidType = FluidRegistry.getFluid(fluidID);
			this.start = start;

			for (int i = 0; i < 6; i++)
			{
				ForgeDirection dir = ForgeDirection.getOrientation(i);

				if (dir == TileGrate.this.getDirection())
				{
					Vector3 check = start.clone().translate(dir);
					this.navigationMap.put(check, start);
					this.workingNodes.add(new TileGrate.ComparableVector(check, 0));
				}
			}
		}

		/**
		 * Tries to fill.
		 * 
		 * @param amount
		 * @param tries
		 * @return Amount filled.
		 */
		public int tryFill(int amount, int tries)
		{
			int filled = 0;

			if (amount >= FluidContainerRegistry.BUCKET_VOLUME)
			{
				for (int i = 0; i < tries; i++)
				{
					ComparableVector next = workingNodes.poll();

					if (next == null)
					{
						TileGrate.this.resetPath();
						return 0;
					}

					if (!isConnected(next.position))
					{
						TileGrate.this.resetPath();
						return 0;
					}

					filled += FluidUtility.fillBlock(TileGrate.this.worldObj, next.position, new FluidStack(fluidType, amount), true);

					if (filled > 0)
					{
						addNextFill(next);
						return filled;
					}
				}
			}

			return filled;
		}

		/**
		 * Adds new nodes into the map.
		 * 
		 * @param next
		 */
		public void addNextFill(ComparableVector next)
		{
			for (int i = 0; i < 6; i++)
			{
				Vector3 newPosition = next.position.clone().translate(ForgeDirection.getOrientation(i));

				if (!this.navigationMap.containsKey(newPosition))
				{
					this.navigationMap.put(newPosition, next.position);
					this.workingNodes.add(new ComparableVector(newPosition, next.iterations + 1));
				}
			}
		}

		public boolean startDrain(Vector3 start)
		{
			fluidType = null;
			this.start = start;

			for (int i = 0; i < 6; i++)
			{
				ForgeDirection dir = ForgeDirection.getOrientation(i);

				if (dir == TileGrate.this.getDirection())
				{
					Vector3 check = start.clone().translate(dir);
					this.navigationMap.put(check, start);
					this.workingNodes.add(new ComparableVector(check, 0));
					fluidType = FluidUtility.getFluidFromBlock(TileGrate.this.worldObj, check);
				}
			}

			return fluidType != null;
		}

		/**
		 * Creates a map of all the fluids.
		 * 
		 * @param tries
		 * @return
		 */
		public boolean tryPopulateDrainMap(int tries)
		{
			if (drainNodes.size() >= Integer.MAX_VALUE)
			{
				return true;
			}

			for (int i = 0; i < tries; i++)
			{
				ComparableVector check = workingNodes.poll();

				if (check == null)
					return true;

				Fluid checkFluid = FluidUtility.getFluidFromBlock(TileGrate.this.worldObj, check.position);

				if (checkFluid != null && fluidType.getID() == checkFluid.getID())
				{
					addNextDrain(check);

					int checkAmount = FluidUtility.getFluidAmountFromBlock(TileGrate.this.worldObj, check.position);

					if (checkAmount > 0)
						this.drainNodes.add(check);
				}
			}

			return false;
		}

		public void addNextDrain(ComparableVector next)
		{
			for (int i = 0; i < 6; i++)
			{
				Vector3 check = next.position.clone().translate(ForgeDirection.getOrientation(i));
				Fluid checkFluid = FluidUtility.getFluidFromBlock(TileGrate.this.worldObj, check);

				if (checkFluid != null && fluidType.getID() == checkFluid.getID())
				{
					if (!navigationMap.containsKey(check))
					{
						navigationMap.put(check, next.position);
						workingNodes.add(new TileGrate.ComparableVector(check, next.iterations + 1));
					}
				}
			}
		}

		/**
		 * Tries to drain a specific amount of fluid.
		 * 
		 * @return - The amount drained.
		 */
		public FluidStack tryDrain(int targetAmount, boolean doDrain)
		{
			int drainedAmount = 0;

			while (!drainNodes.isEmpty())
			{
				ComparableVector fluidCoord = drainNodes.peek();

				if (!isConnected(fluidCoord.position))
				{
					TileGrate.this.resetPath();
					return new FluidStack(fluidType, drainedAmount);
				}

				if (FluidUtility.getFluidFromBlock(TileGrate.this.worldObj, fluidCoord.position) == null || this.fluidType.getID() != FluidUtility.getFluidFromBlock(TileGrate.this.worldObj, fluidCoord.position).getID())
				{
					this.drainNodes.poll();
				}
				else
				{
					int checkAmount = FluidUtility.getFluidAmountFromBlock(TileGrate.this.worldObj, fluidCoord.position);

					if (drainedAmount + checkAmount > targetAmount)
						break;

					if (checkAmount == 0)
					{
						this.drainNodes.poll();
					}
					else
					{
						FluidStack fluidStack = FluidUtility.drainBlock(TileGrate.this.worldObj, fluidCoord.position, doDrain);

						this.drainNodes.poll();

						if (fluidStack != null)
						{
							drainedAmount += fluidStack.amount;

							if (drainedAmount >= targetAmount)
							{
								break;
							}
						}
					}
				}
			}

			TileGrate.this.resetPath();

			if (drainedAmount > 0)
				return new FluidStack(fluidType, drainedAmount);

			return null;
		}

	}

	public static class ComparableVector implements Comparable
	{
		public Vector3 position;
		public int iterations;

		public ComparableVector(Vector3 position, int iterations)
		{
			this.position = position;
			this.iterations = iterations;
		}

		@Override
		public int compareTo(Object obj)
		{
			ComparableVector wr = (ComparableVector) obj;
			if (this.position.y == wr.position.y)
				return this.iterations - wr.iterations;
			return this.position.intY() - wr.position.intY();
		}
	}
}

File: src/main/scala/resonantinduction/archaic/fluid/grate/TileGrate.scala
package resonantinduction.archaic.fluid.grate

import java.util.{Collections, Comparator, HashMap, PriorityQueue}

import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.block.material.Material
import net.minecraft.client.renderer.texture.IIconRegister
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.{ChatComponentText, IIcon}
import net.minecraft.world.IBlockAccess
import net.minecraftforge.common.util.ForgeDirection
import net.minecraftforge.fluids.{Fluid, FluidContainerRegistry, FluidRegistry, FluidStack}
import resonant.lib.mod.config.Config
import resonant.lib.prefab.tile.traits.TRotatable
import resonant.lib.transform.vector.Vector3
import resonant.lib.utility.FluidUtility
import resonantinduction.archaic.fluid.grate.TileGrate._
import resonantinduction.core.Reference
import resonantinduction.core.prefab.node.{NodeFluidPressure, TileFluidProvider}

object TileGrate
{
  @Config(comment = "The multiplier for the influence of the grate. Dependent on pressure.")
  private var grateEffectMultiplier: Double = 5

  @Config(comment = "The speed in which the grate drains blocks. Dependent on grate block influence.")
  private var grateDrainSpeedMultiplier: Double = 0.01

  @SideOnly(Side.CLIENT)
  private var iconFront: IIcon = _

  @SideOnly(Side.CLIENT)
  private var iconSide: IIcon = _
}

class TileGrate extends TileFluidProvider(Material.rock) with TRotatable
{
  fluidNode = new NodeFluidPressure(this)
  private var gratePath: GratePathfinder = _
  private var fillOver: Boolean = true
  isOpaqueCube = false
  normalRender = true

  override def getIcon(world: IBlockAccess, side: Int): IIcon =
  {
    if (side == getDirection.ordinal()) iconFront else iconSide
  }

  override def getIcon(side: Int, metadata: Int): IIcon = if (side == 1) iconFront else iconSide

  @SideOnly(Side.CLIENT)
  override def registerIcons(iconRegister: IIconRegister)
  {
    iconFront = iconRegister.registerIcon(Reference.prefix + "grate_front")
    iconSide = iconRegister.registerIcon(Reference.prefix + "grate")
  }

  override def readFromNBT(nbt: NBTTagCompound)
  {
    super.readFromNBT(nbt)
    fillOver = nbt.getBoolean("fillOver")
  }

  override def writeToNBT(nbt: NBTTagCompound)
  {
    super.writeToNBT(nbt)
    nbt.setBoolean("fillOver", fillOver)
  }

  override def canFill(from: ForgeDirection, fluid: Fluid): Boolean = getDirection != from

  override def canDrain(from: ForgeDirection, fluid: Fluid): Boolean = getDirection != from

  override def update()
  {
    super.update()

    if (!world.isRemote)
    {
      if (ticks % 10 == 0)
      {
        val pressure = fluidNode.asInstanceOf[NodeFluidPressure].pressure(getDirection)
        val blockEffect = Math.abs(pressure * grateEffectMultiplier).toInt
        fluidNode.getPrimaryTank.setCapacity(Math.max(blockEffect * FluidContainerRegistry.BUCKET_VOLUME * grateDrainSpeedMultiplier, FluidContainerRegistry.BUCKET_VOLUME).toInt)

        if (pressure > 0)
        {
          //Output fluid
          if (fluidNode.getFluidAmount >= FluidContainerRegistry.BUCKET_VOLUME)
          {
            if (gratePath == null)
            {
              gratePath = new GratePathfinder(true)
              gratePath.startFill(toVectorWorld, fluidNode.getFluid.getFluid.getID)
            }
            val filledInWorld = gratePath.tryFill(fluidNode.getFluidAmount, blockEffect)
            fluidNode.drain(filledInWorld, true)
          }
        }
        else if (pressure < 0)
        {
          //Input fluid
          val maxDrain = fluidNode.getCapacity - fluidNode.getFluidAmount
          if (maxDrain > 0)
          {
            if (gratePath == null)
            {
              gratePath = new GratePathfinder(false)
              if (!gratePath.startDrain(toVector3))
              {
                resetPath()
              }
            }
            if (gratePath != null && gratePath.tryPopulateDrainMap(blockEffect))
            {
              fluidNode.fill(gratePath.tryDrain(maxDrain, true), true)
            }
          }
        }
      }
    }
  }

  def resetPath()
  {
    this.gratePath = null
  }

  override def fill(from: ForgeDirection, resource: FluidStack, doFill: Boolean): Int = fluidNode.fill(resource, doFill)

  override def drain(from: ForgeDirection, resource: FluidStack, doDrain: Boolean): FluidStack =
  {
    if (resource != null)
    {
      return drain(from, resource.amount, doDrain)
    }
    null
  }

  override def drain(from: ForgeDirection, maxDrain: Int, doDrain: Boolean): FluidStack =
  {
    fluidNode.drain(maxDrain, doDrain)
  }

  protected override def configure(player: EntityPlayer, side: Int, hit: Vector3): Boolean =
  {
    if (!player.isSneaking)
    {
      if (!world.isRemote)
      {
        fillOver = !fillOver
        player.addChatMessage(new ChatComponentText("Grate now set to " + (if (fillOver) "" else "not ") + "fill higher than itself."))
        gratePath = null
      }
      return true
    }
    super.configure(player, side, hit)
  }

  class GratePathfinder(checkVertical: Boolean)
  {
    var fluidType: Fluid = _
    var start: Vector3 = _
    var navigationMap: HashMap[Vector3, Vector3] = new HashMap[Vector3, Vector3]()

    var workingNodes: PriorityQueue[ComparableVector] =
      if (checkVertical) new PriorityQueue[ComparableVector]()
      else new PriorityQueue[ComparableVector](1024, new Comparator[ComparableVector]()
      {
        override def compare(a: ComparableVector, b: ComparableVector): Int =
        {
          val wa = a.asInstanceOf[ComparableVector]
          val wb = b.asInstanceOf[ComparableVector]
          wa.iterations - wb.iterations
        }
      })

    var drainNodes = new PriorityQueue[ComparableVector](1024, Collections.reverseOrder())

    def startFill(start: Vector3, fluidID: Int)
    {
      this.fluidType = FluidRegistry.getFluid(fluidID)
      this.start = start

      val check = start + TileGrate.this.getDirection
      navigationMap.put(check, start)
      workingNodes.add(new ComparableVector(check, 0))
    }

    def tryFill(amount: Int, tries: Int): Int =
    {
      var filled = 0
      if (amount >= FluidContainerRegistry.BUCKET_VOLUME)
      {
        for (i <- 0 until tries)
        {
          val next = workingNodes.poll()
          if (next == null)
          {
            TileGrate.this.resetPath()
            return 0
          }
          if (!isConnected(next.position))
          {
            TileGrate.this.resetPath()
            return 0
          }

          val didFill = FluidUtility.fillBlock(TileGrate.this.worldObj, next.position, new FluidStack(fluidType, amount), true)
          filled += didFill
          if (FluidUtility.getFluidAmountFromBlock(TileGrate.this.worldObj, next.position) > 0 || didFill > 0)
          {
            addNextFill(next)
          }
          if (filled >= amount)
          {
            return filled
          }
        }
      }
      filled
    }

    def addNextFill(next: ComparableVector)
    {
      for (i <- 0 until 6)
      {
        val newPosition = next.position.clone().add(ForgeDirection.getOrientation(i))
        if (!this.navigationMap.containsKey(newPosition) && !(!fillOver && newPosition.yi > y))
        {
          this.navigationMap.put(newPosition, next.position)
          this.workingNodes.add(new ComparableVector(newPosition, next.iterations + 1))
        }
      }
    }

    def startDrain(start: Vector3): Boolean =
    {
      fluidType = null
      this.start = start
      val check = start + TileGrate.this.getDirection
      navigationMap.put(check, start)
      workingNodes.add(new ComparableVector(check, 0))
      fluidType = FluidUtility.getFluidFromBlock(TileGrate.this.worldObj, check)

      fluidType != null
    }

    def tryPopulateDrainMap(tries: Int): Boolean =
    {
      for (i <- 0 until tries)
      {
        val check = workingNodes.poll()
        if (check == null)
        {
          return true
        }
        val checkFluid = FluidUtility.getFluidFromBlock(TileGrate.this.worldObj, check.position)
        if (checkFluid != null && fluidType.getID == checkFluid.getID)
        {
          addNextDrain(check)
          val checkAmount = FluidUtility.getFluidAmountFromBlock(TileGrate.this.worldObj, check.position)
          if (checkAmount > 0)
          {
            drainNodes.add(check)
          }
        }
      }
      return drainNodes.size > 0
    }

    def addNextDrain(next: ComparableVector)
    {
      for (i <- 0 until 6)
      {
        val check = next.position.clone().add(ForgeDirection.getOrientation(i))
        val checkFluid = FluidUtility.getFluidFromBlock(TileGrate.this.worldObj, check)
        if (checkFluid != null && fluidType.getID == checkFluid.getID)
        {
          if (!navigationMap.containsKey(check))
          {
            navigationMap.put(check, next.position)
            workingNodes.add(new ComparableVector(check, next.iterations + 1))
          }
        }
      }
    }

    def tryDrain(targetAmount: Int, doDrain: Boolean): FluidStack =
    {
      var drainedAmount = 0
      var break = false

      //Check all the potential drain coordinates
      while (!drainNodes.isEmpty && !break)
      {
        val fluidCoord = drainNodes.peek()

        if (!isConnected(fluidCoord.position))
        {
          //The coordinate is not connected with the starting coordinate, therefore, we will not drain this.
          TileGrate.this.resetPath()
          return new FluidStack(fluidType, drainedAmount)
        }

        if (FluidUtility.getFluidFromBlock(TileGrate.this.worldObj, fluidCoord.position) == null || this.fluidType.getID != FluidUtility.getFluidFromBlock(TileGrate.this.worldObj, fluidCoord.position).getID)
        {
          drainNodes.poll()
        }
        else
        {
          val checkAmount = FluidUtility.getFluidAmountFromBlock(TileGrate.this.worldObj, fluidCoord.position)

          if (drainedAmount + checkAmount <= targetAmount)
          {
            if (checkAmount == 0)
            {
              drainNodes.poll()
            }
            else
            {
              val fluidStack = FluidUtility.drainBlock(TileGrate.this.worldObj, fluidCoord.position, doDrain, 2)
              //TODO: Do a delayed block update!
              drainNodes.poll()
              if (fluidStack != null)
              {
                drainedAmount += fluidStack.amount
                if (drainedAmount >= targetAmount)
                {
                  break = true
                }
              }
            }
          }
          else
          {
            break = true
          }
        }
      }
      TileGrate.this.resetPath()
      if (drainedAmount > 0)
      {
        return new FluidStack(fluidType, drainedAmount)
      }
      null
    }

    def isConnected(nCheck: Vector3): Boolean =
    {
      var check = nCheck

      if (check.equals(start))
        return true

      //Trace back in the navigation map to see if it reaches the starting point
      do
      {
        check = navigationMap.get(check)

        if (check.equals(null))
          return false

        if (check.equals(start))
          return true
      }
      while (FluidUtility.getFluidFromBlock(TileGrate.this.worldObj, check) != null && fluidType.getID == FluidUtility.getFluidFromBlock(TileGrate.this.worldObj, check).getID)

      return false
    }
  }

  class ComparableVector(var position: Vector3, var iterations: Int) extends Comparable[ComparableVector]
  {
    override def compareTo(obj: ComparableVector): Int =
    {
      val wr = obj.asInstanceOf[ComparableVector]
      if (this.position.y == wr.position.y)
      {
        return this.iterations - wr.iterations
      }
      this.position.yi - wr.position.yi
    }
  }

}
